pkg/kubelet/cm/dra/manager.go

```diff
@@ -120,31 +122,34 @@ func (m *ManagerImpl) PrepareResources(pod *v1.Pod) error {
 			// started, so it's OK (actually correct) to not include it
 			// in the cache.
 			claimInfo.addPodReference(pod.UID)
-			continue
-		}
 
-		// Grab the allocation.resourceHandles. If there are no
-		// allocation.resourceHandles, create a single resourceHandle with no
-		// content. This will trigger processing of this claim by a single
-		// kubelet plugin whose name matches resourceClaim.Status.DriverName.
-		resourceHandles := resourceClaim.Status.Allocation.ResourceHandles
-		if len(resourceHandles) == 0 {
-			resourceHandles = make([]resourcev1alpha2.ResourceHandle, 1)
-		}
+			if claimInfo.preparedByManager {
+				continue
+			}
+		} else {
+			// Grab the allocation.resourceHandles. If there are no
+			// allocation.resourceHandles, create a single resourceHandle with no
+			// content. This will trigger processing of this claim by a single
+			// kubelet plugin whose name matches resourceClaim.Status.DriverName.
+			resourceHandles := resourceClaim.Status.Allocation.ResourceHandles
+			if len(resourceHandles) == 0 {
+				resourceHandles = make([]resourcev1alpha2.ResourceHandle, 1)
+			}
```

Comment:
Something feel odd about this -- it may be correct, but the logic is not very obivous and I'm getting a little lost trying to understand it.

Previously the logic was:
* If the claim is in the cache, we know it has been prepared successfully, so just add a pod reference and continue
* If the claim is not in the cache, attempt to prepare it, and if that succeeds, add it to the cache.

I think the new logic should be:
* If the claim is in the cache and it is prepared, add a pod reference and continue
* If the claim is in the cache, and not prepared, prepare it and mark it as such
* If the claim is not in the cache, prepare it, add it to the cache, and make it as prepared

If you agree, can you update the logic to make this flow more obvious.

Reply:
i agree at high level, however the lower level details are a bit different.

one thing to note is that checking if a claim needs to be processed  (prepared) by the manager and marking it as prepared then actually adding to cache happens in a separate loops

the method has two main loops:
1. go over all pod claims, check them and determine which claims should process by calling to DRA plugin  (filling in `claimInfos` map and `batches`)
2. call DRA plugins in a batch manner and update cache

the changes i introduced:
in 1 :  claims that have not been processed (not prepared) by manager will get added to the relevant structs to be processed
in 2: after a claim has been successfully handled by DRA plugin mark it as prepared before adding it to cache.

lemme try to simplify 1 a bit.